pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,314 @@
+	// First Check Frame Meta Type
+
+	var inputType data.FrameType
+	if frames[0].Meta != nil && frames[0].Meta.Type != "" {
+		inputType = frames[0].Meta.Type
+	}
+	if inputType != "" && !supportedToLongConversion(inputType) {
+		return frames, fmt.Errorf("unsupported input type %s for SQL expression", inputType)
+	}
+
+	// TODO: Add some guessing of Type if not declared
+
+	if inputType == "" {
+		return frames, fmt.Errorf("could not determine input type")
+	}
+
+	convert := getToLongConversionFunc(inputType)
+	if convert == nil {
+		return frames, fmt.Errorf("could not get conversion function for input type %s", inputType)
+	}
+
+	return convert(frames)
+}
+
+func convertNumericMultiToNumericLong(frames data.Frames) (data.Frames, error) {
+	// Apart from metadata, NumericMulti is basically NumericWide, except one frame per thing
+	// so we collapse into wide and call the wide conversion
+	wide := convertNumericMultiToNumericWide(frames)
+	return convertNumericWideToNumericLong(wide)
+}
+
+func convertNumericMultiToNumericWide(frames data.Frames) data.Frames {
+	newFrame := data.NewFrame("")
+	for _, frame := range frames {
+		for _, field := range frame.Fields {
+			if !field.Type().Numeric() {
+				continue
+			}
+			newField := data.NewFieldFromFieldType(field.Type(), field.Len())
+			newField.Name = field.Name
+			newField.Labels = field.Labels.Copy()
+			for i := 0; i < field.Len(); i++ {
+				newField.Set(i, field.CopyAt(i))
+			}
+			newFrame.Fields = append(newFrame.Fields, newField)
+		}
+	}
+	return data.Frames{newFrame}
+}
+
+func convertNumericWideToNumericLong(frames data.Frames) (data.Frames, error) {
+	// Wide should only be one frame
+	if len(frames) != 1 {
+		return nil, fmt.Errorf("expected exactly one frame for wide format, but got %d", len(frames))
+	}
+	inputFrame := frames[0]
+
+	// The Frame should have no more than one row
+	if inputFrame.Rows() > 1 {
+		return nil, fmt.Errorf("expected no more than one row in the frame, but got %d", inputFrame.Rows())
+	}
+
+	// Gather Unique Label Keys from each numeric, and unique numeric field names,
+	// each one maps to a field in the output long Frame.
+	uniqueNames := make([]string, 0)
+	uniqueKeys := make([]string, 0)
+
+	uniqueNamesMap := make(map[string]data.FieldType)
+	uniqueKeysMap := make(map[string]struct{})
+
+	prints := make(map[string]int)
+
+	registerPrint := func(labels data.Labels) {
+		fp := labels.Fingerprint().String()
+		if _, ok := prints[fp]; !ok {
+			prints[fp] = len(prints)
+		}
+	}
+
+	for _, field := range inputFrame.Fields {
+		if field.Type().Numeric() {
+			if _, ok := uniqueNamesMap[field.Name]; !ok {
+				uniqueNames = append(uniqueNames, field.Name)
+			}
+			uniqueNamesMap[field.Name] = field.Type()
+
+			if field.Labels != nil {
+				for key := range field.Labels {
+					if _, ok := uniqueKeysMap[key]; !ok {
+						uniqueKeys = append(uniqueKeys, key)
+					}
+					uniqueKeysMap[key] = struct{}{}
+					registerPrint(field.Labels)
+				}
```

Comment:
Why are we doing this once per label-key? Shouldn't we do this once per `field.Labels`? This doesn't take `key` as an argument.
```suggestion
				}
				registerPrint(field.Labels)
```